File: server/src/models/alert.rs
use chrono::{DateTime, Utc};
use serde::Serialize;

/// 가격 알림 유형
#[derive(Debug, Clone, PartialEq, Eq, Serialize, sqlx::Type)]
#[sqlx(type_name = "TEXT", rename_all = "snake_case")]
pub enum AlertType {
    TargetPrice,
    BelowAverage,
    NearLowest,
    AllTimeLow,
}

/// 카테고리 알림 조건
#[derive(Debug, Clone, PartialEq, Eq, Serialize, sqlx::Type)]
#[sqlx(type_name = "TEXT", rename_all = "snake_case")]
pub enum AlertCondition {
    PriceDropPercent,
    AllTimeLow,
    BuyTiming,
    SalesSpike,
}

/// price_alerts 테이블
#[derive(Debug, Clone, sqlx::FromRow, Serialize)]
pub struct PriceAlert {
    pub id: i64,
    pub user_id: i64,
    pub product_id: i64,
    pub alert_type: AlertType,
    pub target_price: Option<i32>,
    pub is_active: bool,
    pub last_triggered_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// category_alerts 테이블
#[derive(Debug, Clone, sqlx::FromRow, Serialize)]
pub struct CategoryAlert {
    pub id: i64,
    pub user_id: i64,
    pub category_id: i32,
    pub alert_condition: AlertCondition,
    pub threshold_percent: Option<i32>,
    pub max_price: Option<i32>,
    pub is_active: bool,
    pub last_triggered_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// keyword_alerts 테이블
#[derive(Debug, Clone, sqlx::FromRow, Serialize)]
pub struct KeywordAlert {
    pub id: i64,
    pub user_id: i64,
    pub keyword: String,
    pub category_id: Option<i32>,
    pub max_price: Option<i32>,
    pub is_active: bool,
    pub last_triggered_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

File: app/lib/models/alert.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'alert.freezed.dart';
part 'alert.g.dart';

/// 가격 알림 — target_price, below_average, near_lowest, all_time_low
@freezed
abstract class PriceAlert with _$PriceAlert {
  const factory PriceAlert({
    required int id,
    @JsonKey(name: 'user_id') required int userId,
    @JsonKey(name: 'product_id') required int productId,
    @JsonKey(name: 'alert_type') required String alertType,
    @JsonKey(name: 'target_price') int? targetPrice,
    @JsonKey(name: 'is_active') @Default(true) bool isActive,
    @JsonKey(name: 'last_triggered_at') DateTime? lastTriggeredAt,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _PriceAlert;

  factory PriceAlert.fromJson(Map<String, dynamic> json) =>
      _$PriceAlertFromJson(json);
}

/// 카테고리 알림
@freezed
abstract class CategoryAlert with _$CategoryAlert {
  const factory CategoryAlert({
    required int id,
    @JsonKey(name: 'user_id') required int userId,
    @JsonKey(name: 'category_id') required int categoryId,
    @JsonKey(name: 'alert_condition') @Default('any_drop') String alertCondition,
    @JsonKey(name: 'threshold_percent') int? thresholdPercent,
    @JsonKey(name: 'max_price') int? maxPrice,
    @JsonKey(name: 'is_active') @Default(true) bool isActive,
    @JsonKey(name: 'last_triggered_at') DateTime? lastTriggeredAt,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _CategoryAlert;

  factory CategoryAlert.fromJson(Map<String, dynamic> json) =>
      _$CategoryAlertFromJson(json);
}

/// 키워드 알림
@freezed
abstract class KeywordAlert with _$KeywordAlert {
  const factory KeywordAlert({
    required int id,
    @JsonKey(name: 'user_id') required int userId,
    required String keyword,
    @JsonKey(name: 'category_id') int? categoryId,
    @JsonKey(name: 'max_price') int? maxPrice,
    @JsonKey(name: 'is_active') @Default(true) bool isActive,
    @JsonKey(name: 'last_triggered_at') DateTime? lastTriggeredAt,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _KeywordAlert;

  factory KeywordAlert.fromJson(Map<String, dynamic> json) =>
      _$KeywordAlertFromJson(json);
}

/// GET /alerts 응답 — { price_alerts, category_alerts, keyword_alerts }
@freezed
abstract class AlertListResponse with _$AlertListResponse {
  const factory AlertListResponse({
    @JsonKey(name: 'price_alerts') @Default([]) List<PriceAlert> priceAlerts,
    @JsonKey(name: 'category_alerts') @Default([]) List<CategoryAlert> categoryAlerts,
    @JsonKey(name: 'keyword_alerts') @Default([]) List<KeywordAlert> keywordAlerts,
  }) = _AlertListResponse;

  factory AlertListResponse.fromJson(Map<String, dynamic> json) =>
      _$AlertListResponseFromJson(json);
}
